README.md
<h1 align="center">
	<img src="https://msfninja.000webhostapp.com/schedule/logo/logo-transparent-blue.png" alt="Schedule Logo" width="500" />
	<br /><br />
	<p>To-dos, Notes, Calendar and More</p>
</h1>

Demo hidden onion service: [h73ug4anp73u7uzcr5smaamt3ejgqiumpht5oagtvxbmoeea2uz3quad.onion](https://h73ug4anp73u7uzcr5smaamt3ejgqiumpht5oagtvxbmoeea2uz3quad.onion)

**Please note that currently the project is under development. This means none of the code in this repository is guaranteed to be working.**

Schedule is a web application, allowing you to create, save and sync contacts, to-dos, notes, calendars, school information and other weird stuff like your school timetable on a server, and make it accessible, manageable and sync-able from and on any device<sup id="l-ref-1"><a href="#ref-1">1</a></sup>.

It's a valuable tool for any student, or generally, anyone who necessitates keeping track of his tasks in the forms of to-os or store information in the forms of memoranda, or even an agenda. Since Schedule is a web application, inferring it is accessed from a web browser<sup id="l-ref-2"><a href="#ref-2">2</a></sup>, it can be used on implicitly any device with a web browser and access to the internet or network on which the server hosting Schedule is running. For accessibility (specifically on mobile devices), Schedule allows a progressive web application that you can install as a (browser emulated) application, which provides access to it with ease right from your device's home screen (or your device-specific environment).

Schedule is simple to use, has a fine and candid graphical user interface and works seamlessly. Apart from that, what I'd like to denote is that Schedule is quite privacy-friendly, as it stores all user data<sup id="l-ref-3"><a href="#ref-3">3</a></sup> on the server encrypted, using the Advanced Encryption Standard algorithm with a 256-bit key size, which assures nobody but you can access your data.

<table>
	<tr>
		<th>Table of Contents</th>
	</tr>
	<tr>
		<td>
			<ol>
				<li><a href="#notice">Notice</a></li>
				<li><a href="#prerequisites">Prerequisites</a></li>
				<ol>
					<li><a href="#nodejs-installation">Node.js Installation</a></li>
					<li><a href="#npm-modules">Npm Modules</a></li>
					<ol>
						<li><a href="#nodemon-installation">Nodemon Installation</a></li>
						<li><a href="#sass-installation">Sass Installation</a></li>
						<li><a href="#coffeescript-installation">CoffeeScript Installation</a></li>
						<li><a href="#other-modules">Other Modules</a></li>
					</ol>
					<li><a href="#git-installation">Git Installation</a></li>
				</ol>
				<li><a href="#repository-cloning">Repository Cloning</a></li>
				<li><a href="#initial-configuration">Initial Configuration</a></li>
				<li><a href="#server">Server</a></li>
				<ol>
					<li><a href="#ssl-certificate">SSL Certificate</a></li>
					<li><a href="#server-initiation">Server Initiation</a></li>
					<li><a href="#server-configuration">Server Configuration</a></li>
					<li><a href="#compiling-sass-code">Compiling Sass Code</a></li>
					<li><a href="#compiling-coffeescript-code">Compiling CoffeeScript Code</a></li>
				</ol>
				<li><a href="#support">Support</a></li>
				<li><a href="#notes">Notes</a></li>
			</ol>
		</td>
	</tr>
</table>

## Notice

Schedule is an open-source project licensed under the [GNU General Public License version 3.0](https://www.gnu.org/licenses/gpl-3.0.html). To see what you may and may not do with the source code of this project, see the [license](https://github.com/msfninja/schedule/blob/main/LICENSE). The license can also be found in the root directory of the project as `LICENSE`.

Schedule is open-source because 1) this way users can see how the utility works and does for themselves and 2) other developers can utilize the code.

This `README.md` file will be assuming you are using a Linux distribution, so when encountering commands or other system-related things and using an operating system other than a Linux-based one or a Linux distribution that is not featured, make sure you use your system's equivalent alternative or approach of achieving the outcome.

## Prerequisites

If you want to host Schedule yourself on your server, you'd need several items present for it to be possible. Not all items are mandatory, but some may be found useful.

### Node.js Installation

Frist of all, you'll need [Node.js](https://nodejs.org/en/). Schedule essentially boils down to a node application, so to run the application, you need Node.js. You can download an installer or the source code from their [official website](https://nodejs.org/en/download/), or run these commands in your shell:

**Ubuntu/Debian:**

```bash
sudo apt -y install nodejs
```

**Fedora:**

```bash
sudo dnf install nodejs
```

**RedHat/CentOS:**

```bash
sudo yum -y install nodejs
```

This should install Node.js on your system.

You can check what version of Node.js is installed (and if Node.js is actually installed) on your system by issuing the following in your shell:

```bash
node --version
```

If this returns an error, or doesn't output anything, it means that either Node.js is not installed, or installed incorrectly. Try reinstalling it in this case.

### Npm Modules

Apart from Node.js, this node application requires several modules to operate properly which do not come preinstalled with Node.js by default. The modules `ip`, `yaml`, `uuid` and `colors` are required for Schedule to work, and other utilities like `nodemon` come in handy, but are optional. Further on there is also Sass and CoffeeScript, but you would only need those if you are planning on modifying any of the `.sass` or `.coffee` files (presuming you are going to compile the code as well to reflect the changes).

#### Nodemon Installation

This one is optional, but really makes your life easier and frees from a batch of headaches. Instead of manually reinitiating your node application upon each a crash of it or modification to its source code, you can automate this process by using [nodemon](https://nodemon.io/). To install nodemon, issue this command in your shell:

```bash
npm i -g nodemon
```

The `-g` option will install nodemon globally, so you can use it system-wide in any project<sup id="l-ref-4"><a href="#ref-4">4</a></sup>. Again, you can check the version of `nodemon` installed or its presence in the first place by running the following command:

```bash
nodemon --version
```

#### Sass Installation

Proceed with this step only if you are planning on modifying Schedule's `.sass` files and compling them, and you don't have it present on your machine yet.

To install Sass on your machine, run this command in your shell:

```bash
npm i -g sass
```

The `-g` option is the same story here as with `nodemon` mentioned [before](#nodemon-installation). You can check what version of Sass you have installed or if it is present by issuing the following command:

```bash
sass --version
```

#### CoffeeScript Installation

Proceed with this step only if you are planning on modifying Schedule's `.coffee` files and compling them, and you don't have it present on your machine yet.

To install CoffeeScript on your machine, issue the following:

```bash
npm i -g coffeescript
```

CoffeeScript's version or presence can be checked by running the following in your shell:

```bash
coffee --version
```

#### Other Modules

**Important: you should locally clone Schedule's repository before proceeding with this section, as you need a place to install the npm modules to. Follow [this section](#repository-cloning) first, and then proceed with installing the mandatory npm modules.**

As mentioned [earlier](#npm-modules), the `ip`, `yaml`, `uuid` and `colors` modules do not come preinstalled with Node.js by default. This means, for Schedule to work properly, you have to install them manually using the (CLI) npm package manager.

You can issue the following command in your shell that will install them all (by default I have the `node_modules` directory inside the `server` directory, hence the `cd schedule/server` (assuming you are in the working directory to which you cloned Schedule's repository to (from [this section](#repository-cloning))) before the actual installation of the npm modules):

```bash
cd schedule/server; npm i ip yaml uuid colors
```

A brief description of what every module is used for:

 - `ip`: the `ip` module is used solely for one purpose&#8212;retrieve the server's internet protocol address. It is not used for any IP address logging or anything similar as it may seem. Schedule does not keep any personal or possbily identifiable user data logs, or any logs at all, and you can verify that in the source code.
 - `yaml`: the `yaml` module is used to parse YAML files. An example can be the `config.yml` file in the root directory of the project, which holds various configuration properties for Schedule that are read by the `server/server.js` file.
 - `uuid`: the `uuid` module is used to generate universally unique identifiers. An example usage is using them as tokens for user sessions and similar activities.
 - `colors`: the `colors` module is really just used to style text output in the console, and that's what it's supposed to be used for.

 If you already read through the [repsitory cloning](#repository-cloning) and [git installation](#git-installation) sections, you can skip to [this section](#initial-configuration).

### Git Installation

To clone Schedule's repository to your machine and use it, you need git. Although you can use other alternatives like GitHub's CLI utility, here I will be using git.

If you don't have git, you can install it by running the following commands in your shell:

**Ubuntu/Debian:**

```bash
sudo apt -y install git
```

**Fedora:**

```bash
sudo dnf install git
```

**RedHat/CentOS:**

```bash
sudo yum -y install git
```

Git's version or presence on the machine can be checked by running the following in your shell:

```bash
git --version
```

Now that you have installed git, you can proceed with [this section](#repository-cloning) on cloning Schedule's repository.

## Repository Cloning

Before you start cloning using git, make sure you have git present on your machine. If you don't have git, proceed with [this section](#git-installation) first. Otherwise, make sure you are in your desired working directory (in your shell). If you want to clone Schedule to the your home directory, issue the following to change your current working directory to your home directory in your shell:

```bash
cd ~
```

To clone Schedule's repository, run the following command in your shell:

```bash
git clone https://github.com/msfninja/schedule.git
```

This should clone Schedule's repository to your current working directory. You can check if it's cloned by listing your current working directory contents that contain `schedule` by running the following:

```bash
ls -al|grep schedule
```

Or you can simply try to go to the cloned directory by running the following in your shell:

```bash
cd schedule
```

After successfully cloning Schedule's repository to your machine, you can proceed with installing other necessary npm modules in [this section](#other-modules).

## Initial Configuration

Before you can actually run the server (`server/server.js` file), you need to run the `configure` file to set the project up.

To run `configure`, change your current working directory to the project's root directory, and run the following in your shell:

```bash
./configure
```

This will create the `schedule-data` directory in your current user home directory with additional subdirectories that will later store user data, root account data, various authentication data and public data to be displayed in the web client of Schedule.

**Important: make sure you will be running the node server as the same user you were running the `configure` file, otherwise the server won't run properly by reading a different home directory than the `configure` file created the `schedule-data` directory in.**

If you're here from the server initiation section, you can skip [to that section](#server-initiation).

## Server

There are a few more things you have to do before running the `server/server.js` file. Schedule has several features which require its traffic to go over the hypertext transfer protocol **secure**, rather than HTTP. First of all, passwords and sensitive information alike are being transmitted over the network between the client and the server. Additionally, Schedule offers a PWA functionality, which by the majority of browsers I know has a requirement of its traffic to be transmitted over HTTPS.

### SSL Certificate

If you have a secure sockets layer certificate, put it and the rest that belongs to it into the `server/ssl` directory, and rename the key file to `key.pem`, and the certificate file to `cert.pem`.

To generate a self-signed SSL certificate using `openssl`, go to the `server/ssl` directory:

```bash
cd server/ssl
```

And run the following commands in your shell:

```bash
openssl genrsa -out key.pem
openssl req -new -key key.pem -out csr.pem
openssl x509 -req -days 9999 -in csr.pem -signkey key.pem -out cert.pem
rm csr.pem
```

This will leave you with two files: `cert.pem` and `key.pem` (inside the `server/ssl` directory). That is all you need to do. The server file will utilize these files.

### Server Initiation

**Important: before you can run the server, you need to run the `configure` file. Otherwise the node application will exit with an error. Go to [this section](#initial-configuration) to get started with the `configure` file.**

If you've run the `configure` file and are ready to go, you can initialize the server. Go to the `server` directory:

```bash
cd server
```

And initiate the node server by running this command in your shell:

```bash
nodemon server.js --ignore '*.json'
```

The `--ignore` option is set to `'*.json'` so the server won't reinitiate every time there is a change to any JSON file within the `server` directory.

### Server Configuration

Server configuration is not very broad, but present nevertheless. Common hand-picked values that are used throughout the `server/server.js` file are collected in the `config.yml` file in the root directory of the project. If you need to change something miscellaneous, you can try to find the property there.

For example, if you'd like to change the port, you can search for these lines:

```yaml
server:
  port: 150
```

Where presumably the default number 150 represents the port. Changing this to a value that is not a valid transmission control protocol or user datagram protocol port will result in an error upon initiating the server.

### Compiling Sass Code

Make sure you have Sass installed before proceeding with this section. To see how to install Sass, follow [this section](#sass-installation) first.

So far, Schedule has only one Sass style sheet. To compile it into CSS code that will be served to the client, run the below command in your shell (assuming your current working directory is the project's root directory):

```bash
sass ./server/client/sass/stylesheet.sass ./public/client/css/stylesheet.css
```

This will compile the code only once though. If you want to automate the compilation process, you can add the `-w` option like this:

```bash
sass -w ./server/client/sass/stylesheet.sass ./public/client/stylesheet.css
```

By default, Sass will also make a source map along the compiled CSS style sheet. If you don't want a source map, you can add the `--no-source-map` option.

### Compiling CoffeeScript Code

Make sure you have CoffeeScript installed before proceeding with this section. To see how to install CoffeeScript, follow [this section](#coffeescript-installation) first.

Schedule has 2 CoffeeScript scripts, which are `server/client/coffee/main.coffee` and `server/client/coffee/user.coffee`. Both can be compiled into JavaScript code that will be served to the client issuing the following two commands for both scripts in your shell respectively:

```bash
coffee -o ./public/client/js/main.js -c ./server/client/coffee/main.coffee
```

```bash
coffee -o ./public/client/js/user.js -c ./server/client/coffee/user.coffee
```

CHANGELOG
Unreleased 2021-09-16 msfninja <msfninja@airmail.cc>:
	
	Note:
	Initiation of project. Summing up all development from its creation to the date of this release.

	Added:
	 - Client
	   * A web application accessible from any device (featuring a PWA) to interact with your data
	 - Server
	   * A server that handles the general thing
	     + A HTTP server for handling the web client
	     + A WebDAV server for remote configuration
	 - CLI
	   * A CLI (server-side) that can perform numerous tasks

server/server.js
// Schedule node app server source code. This file was
// written by msfninja <msfninja@airmail.cc>.
//
// (c) 2021 Schedule

'use strict';

const // modules
	https = require('https'),
	path = require('path'),
	url = require('url'),
	fs = require('fs'),
	qs = require('querystring'),
	ip = require('ip'),
	crypto = require('crypto'),
	bcrypt = require('bcrypt'),
	{
		v1: uuidv1,
		v4: uuidv4
	} = require('uuid'),
	YAML = require('yaml'),
	colors = require('colors'),
	os = require('os'),
	{ MongoClient } = require('mongodb');

const // cli object constructor (global scope)
	cli = new CLI();

const // repo
	repo = {
		gh: {
			url: 'https://github.com/msfninja/schedule',
			blob: 'https://github.com/msfninja/schedule/blob/main'
		}
	};

let // config and db
	config,
	db;

try {
	config = YAML.parse(
		fs.readFileSync(
			path.join(path.resolve(__dirname,'..'),'config.yml')
		).toString());

	if (!config) throw new Error('The config.yml file seems to be incorrectly configured.');
}
catch (e) {
	cli.err(
		e,
		`Download the proper config.yml file from the repository:\n\n${repo.gh.blob}/config.yml`
	);
	process.exit(1);
}

const // dirs
	dir = path.resolve(__dirname,'..'),
	dirs = {
		usr: os.homedir(),
		dat: path.join(os.homedir(),`${config.app.alt_name}-data`)
	};

const // mongodb
	dburl = 'mongodb://127.0.0.1:27017',
	dbclient = new MongoClient(dburl);

const // ssl
	opt = {
		key: fs.readFileSync(
			path.join(dir,'server/ssl/key.pem')
		),
		cert: fs.readFileSync(
			path.join(dir,'server/ssl/cert.pem')
		)
	};

const // basic operations
	ex = h => {
		return fs.existsSync(h);
	},
	rd = h => {
		return ex(h) ? fs.readFileSync(h).toString() : '';
	},
	compare = (a,b) => {
		let arr = [];
		for (var i = 0; i < a.length; i++) for (var j = 0; j < b.length; j++) if (a[i] === b[j]) arr.push(a[i]);
		return {
			array: arr,
			match: arr.length > 0,
			size: arr.length
		};
	},
	format = s => {
		return s.toLowerCase().replace(/[^a-z0-9\-]/g,'').replace(/[\t \n]/g,'-'); 
	};

const // app functions
	term = (r,s,h,c) => {
		r.writeHead(s,h);
		if (c) r.write(c);
		r.end();
	},
	render = (res,req,h,m) => {
		let
			html = rd(h),
			modes = m.split(''),
			accept = ['a','r','u','x'],
			components = [
				[
					'doctype',
					'footer',
					'header',
					'link',
					'loader',
					'meta',
					'noscript',
					'script'
				],
				[
					'usr.nav',
					'usr.pwa',
					'usr.script'
				]
			];

		if (compare(modes,accept).match) {
			if (modes.includes(accept[accept.indexOf('r')])) {
				for (var i = components[0].length - 1; i >= 0; i--) {
					html = html.replace(
						new RegExp(`\{component\.${components[0][i]}\}`,'g'),
						rd(
							path.join(dir,`server/client/component/${components[0][i]}.xhtml`)
						)
					);
				}

				for (const prop in config.app) {
					if (typeof config[prop] !== 'object') {
						html = html.replace(
							new RegExp(`\{app\.${prop}\}`,'g'),
							config.app[prop]
						);
					}
				}
			}
			if (modes.includes(accept[accept.indexOf('u')])) {
				let
					user = new User(res,req),
					cookie = cookies(req),
					dat = user.request('tkn',cookie['UTOKEN']);

				for (var i = components[1].length - 1; i <= 0; i--) {
					html = html.replace(
						new RegExp(`\{component\.${components[1][i]}\}`,'g'),
						rd(
							path.join(dir,`server/client/component/${components[1][i]}.xhtml`)
						)
					);
				}

				html = html
					.replace(/\{nav.tabs.true\}/g,createNav(true))
					.replace(/\{nav.tabs.false\}/g,createNav(false))
					.replace(/\{user\.name\}/g,dat.usr);
			}
		}

		return html;
	};

const // site functions
	cookies = r => {
		var
			obj = {},
			rc = r.headers.cookie;
		rc && rc.split(';').forEach(e => {
			var parts = e.split('=');
			obj[parts.shift().trim()] = decodeURIComponent(parts.join('='));
		});
		return obj;
	},
	createNav = c => {
		let
			html = '',
			navdat = YAML.parse(rd(`${dir}/app/nav.yml`)).nav;

		navdat.forEach(e => {
			if (c) {
				if (e.nav) {
					html += `<div title="Go to ${e.name}" id="btn-nav-${format(e.name)}" class="nav-btn" onclick="window.location.href = '${e.href}';">
							<div class="ico">
								<span><i class="bi bi-${e.icon}"></i></span>
							</div>
							<div class="txt">
								<span>${e.name}</span>
							</div>
						</div>`;
				}
			}
			else {
				if (!e.nav) {
					html += `<div title="Go to ${e.name}" id="btn-${format(e.name)}" class="btn" onclick="window.location.href = '${e.href}';">
							<div class="ico">
								<span><i class="bi bi-${e.icon}"></i></span>
							</div>
							<div class="txt">
								<span>${e.name}</span>
							</div>
						</div>`;
				}
			}
		});
		return html;
	};

function CLI() {
	this.log = s => {
		console.log(`${config.cli.name.blue.bold}> ${s}`);
	};

	this.err = (e,s) => {
		console.clear();
		console.error(`${'\nERROR:'.red.bold}\n\n${e}\n`);
		if (s) console.log(`${'SOLUTION:'.blue.bold}\n\n${s}\n`);
	};

	this.cls = () => {
		console.clear();
	};
}

function Root(res,req) { // root function // NOT READY FOR USE
	let rtokens = new RTokens(res);

	this.verify = a => { // verify root credentials or token
		if (a) {
			if (Object.prototype.toString.call(a) === '[object Array]') {
				return a[0] === keys('usr') && a[1] === decrypt(keys('psw'),`${a[1]}${guid()}`.substr(0,32));
			}
			else if (typeof a === 'string') {
				return rtokens.verify(a);
			}
		}
		return false;
	};

	this.login = c => { // login for root
		if (c) {
			term(res,200,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/root/panel/index.xhtml`,'arx'));
		}
		else {
			term(res,200,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/root/index.xhtml`,'ar'));
		}
	};

	this.cred = (t,a) => { // credentials recovery // NOT READY FOR USE
		if (t === 'set') {
			let obj = {};
			
			obj.usr = a[0];
			obj.psw = encrypt(
				a[1],
				`${a[1]}${guid()}`.substr(0,32)
			);

			fs.writeFile(`${dirs.dat}/auth/root/dat/keys.hash`,JSON.stringify(obj),err => {
				if (err) throw err;
			});
		}
	};
}

function RTokens(res) { // root tokens function
	this.create = () => {};

	this.verify = t => {
		return false;
	};

	this.delete = t => {};
}

function User(res,req) { // user function
	let utokens = new UTokens(res);

	this.create = o => { // create user account
		let
			hash = bcrypt.hash(o.psw,config.server.security.salt_rounds,(e,h) => {
				if (e) throw e;
				return h;
			}),
			key = crypto.pbkdf2(hash,guuid(),config.server.security.pbkdf2_iter,32,'sha512',(e,k) => {
				if (e) throw e;
				return k;
			}),
			ckey = getuuid(2).substring(0,32),
			obj = {
				usr: o.usr,
				hash: hash,
				key: encrypt(key,ckey),
				ckey: ckey
			},
			h = path.join(dirs.dat,'usr',obj.usr),
			arr = [ path.join(h,'cnt') ];

		o = null;

		try {
			fs.accessSync(
				path.join(dirs.dat,'usr'),
				fs.constants.R_OK | fs.constants.W_OK
			);
		}
		catch (e) {
			try {
				fs.chmodSync(
					path.join(dirs.dat,'usr'),
					0o724
				);
			}
			catch (f) {
				cli.err(
					`${config.app.name} cannot save any user data to this directory:\n\n${h}\n\n${f}`,
					`Check the permissions of the user you run this node app as on that directory.`
				);
				process.exit(1);
			}
		}

		arr.forEach(e => {
			fs.mkdirSync(
				e,
				{ recursive: true }
			);
		});

		fs.writeFile(
			path.join(dirs.dat,'usr',obj.usr,'data.json'),
			JSON.stringify({
				tkn: null
			}),
			e => {
				if (e) throw e;
			}
		);

		db.collection('users').insertOne(obj,e => {
			if (e) throw e;
			utokens.create(obj.usr,true);
		});
	};

	this.verify = a => { // verify user credentials or token
		if (a) {
			if (typeof a === 'object' && !Array.isArray(a)) {
				let
					usr = async () => {
						await db.collection('users').find({},(e,r) => {
							if (e) throw e;
							r.toArray((f,s) => {
								if (f) throw f;
								return s;
							});
						}).find(e => e === a.usr);
					};

				if (usr) {
					let
						obj = {
							usr: usr,
							psw: a.psw
						},
						dat = JSON.parse(
							rd(
								path.join(dirs.dat,'usr',obj.usr,'data.json')
							)
						),
						match = bcrypt.compare(obj.psw,dat.hash,(e,r) => {
							return e ? false : r;
						});

					return dat.usr === obj.usr && match;
				}
			}
			else if (typeof a === 'string') {
				let
					tkn = fs.readdirSync(
						path.join(dirs.dat,'usr')
					).find(e =>
						utokens.verify(
							JSON.parse(
								rd(
									path.join(dirs.dat,'usr',e,'data.json')
								)
							).utoken
						)
					);

				return tkn ? true : false;
			}
		}
		return false;
	};

	this.login = c => { // login for user
		if (c) {
			term(
				res,200,
				{ 'Content-Type': 'application/xhtml+xml' },
				render(
					res,req,
					path.join(dir,'app/index.xhtml'),
					'ru'
				)
			);
		}
		else {
			term(
				res,200,
				{ 'Content-Type': 'application/xhtml+xml' },
				render(
					res,req,
					path.join(dir,'public/index/index.xhtml'),
					'r'
				)
			);
		}
	};

	this.check = u => { // check if user exists
		let
			usr = async () => {
				await db.collection('users').find({},(e,r) => {
					if (e) throw e;
					return r.usr;
				});
			};

		return usr ? true : false;
	};

	this.request = (t,d) => { // request user data by username or token
		let
			usr = async () => {
				await db.collection('users').find({},(e,r) => {
					if (e) throw e;
					r.toArray((f,s) => {
						if (e) throw e;
						return s;
					});
				});
			};

		if (usr) {
			if (t === 'usr') {
				let target = usr.find(e => e === d);

				if (target) return target;
			}
			else if (t === 'tkn') {
				if (utokens.get()) {
					let target = utokens.get().find(e => e.tkn === d);

					if (target) {
						return JSON.parse(
							rd(
								path.join(dirs.dat,'usr',target.usr,'data.json')
							)
						);
					}
				}
			}
		}

		return false;
	};
}

function UTokens(res) { // user tokens function
	let utokens;

	try {
		utokens = JSON.parse(
			rd(
				path.join(dirs.dat,'auth/usr/tokens.json')
			)
		);
	}
	catch (e) {
		utokens = false;
	}

	this.get = () => { // get array of user tokens
		return utokens;
	};

	this.create = (u,c) => { // create user token
		let
			time = new Date().valueOf(),
			expires = config.server.usr.session.time,
			token = getuuid(1),
			arr = [],
			obj = {
				usr: u,
				tkn: token,
				exp: time + expires
			};

		if (utokens) arr = utokens;

		arr.push(obj);

		fs.writeFile(
			path.join(dirs.dat,'auth/usr/tokens.json'),
			JSON.stringify(arr),
			err => {
				if (err) {
					term(
						res,500,
						{ 'Content-Type': 'application/xhtml+xml' },
						render(
							res,req,
							path.join(dir,'server/client/err/500.xhtml'),
							'r'
						)
					);
				}
			}
		);

		obj = JSON.parse(
			rd(
				`${dirs.dat}/usr/${u}/data.json`
			)
		);
		obj.tkn = token;

		fs.writeFile(`${dirs.dat}/usr/${u}/data.json`,JSON.stringify(obj),err => {
			if (err) term(res,500,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/server/client/err/500.xhtml`,'r'));
		});

		let exp = new Date(time + expires).toUTCString();

		if (c) term(res,302,[
			['Set-Cookie',`UTOKEN=${token}; Path=/; Expires=${exp};`],
			['Set-Cookie',`HASH=${obj.key}; Path=/; Expires=${exp};`],
			['Set-Cookie',`KEY=${obj.ckey}; Path=/; Expires=${exp};`],
			['Location','/usr']
		]);
	};

	this.verify = t => { // verify user token
		if (utokens) {
			let tkn = utokens.find(e => e.tkn === t);

			if (tkn) return true;
		}
		return false;
	};

	this.delete = (t,c,d) => { // delete user token from tokens.json and/or user data
		if (utokens) {
			let
				user = new User(res),
				dat = user.request('tkn',t),
				target = utokens.find(e => e.tkn === t);

			if (target) {
				if (c || target.exp > new Date().valueOf()) {
					utokens.splice(utokens.indexOf(target),1);
					fs.writeFile(`${dirs.dat}/auth/usr/tokens.json`,JSON.stringify(utokens),err => {
						if (err) throw err;
					});
				}
				if (d) {
					delete dat.tkn;
					fs.writeFile(`${dirs.dat}/usr/${dat.usr}/data.json`,JSON.stringify(dat),err => {
						if (err) throw err;
					});
				}
			}

			dat = '';
			target = '';
		}
	};
}

function Todos(res) { // to-dos management
	let
		user = new User(res),
		cl = db.collection('todos');

	this.add = (t,o) => {
		if (user.verify(t)) {
			//
		}
	};

	this.get = t => {};

	this.delete = (t,i) => {};
}

function Notes(res) { // notes management // NOT READY FOR USE
	let user = new User(res);

	this.add = (t,o) => {
		if (user.verify(t)) {
			let
				dt = new Date(),
				u = user.request('tkn',t);

			fs.writeFile(`${dirs.dat}/usr/${u.usr}/cnt/notes/${dt.getFullYear()}${dt.getMonth()}${dt.getDate()}${dt.getHours()}${dt.getMinutes()}${dt.getSeconds()}.json`,JSON.stringify(o),err => {
				if (err) throw err;
			});
		}
	};

	this.get = t => {
		if (user.verify(t)) {
			let
				arr = [],
				u = user.request('tkn',t);

			fs.readdirSync(`${dirs.dat}/usr/${u.usr}/cnt/notes`).forEach(e => {
				let todo = JSON.parse(rd(`${dirs.dat}/usr/${u.usr}/cnt/notes/${e}`));
				todo.id = e;
				arr.push(todo);
			});

			return arr;
		}
	};

	this.delete = (t,i) => {
		if (user.verify(t)) {
			let
				u = user.request('tkn',t),
				target = fs.readdirSync(`${dirs.dat}/usr/${u.usr}/cnt/notes`).find(e => e === i);

			if (target) {
				fs.unlink(`${dirs.dat}/usr/${u.usr}/cnt/notes/${target}`,err => {
					if (err) throw err;
				})
			}
		}
	};
}

function Calendar(res) { // calendar management
	//
}

function School() { // schools/timetables management
	//
}

const // https server
	start = port => {
		https.createServer(opt,(req,res) => {
			let
				q = url.parse(req.url,true),
				p = q.pathname,
				ip = req.connection.remoteAddress.replace(/^\:+[a-z]{4}\:/i),
				cookie = cookies(req),
				root = new Root(res,req),
				rtokens = new RTokens(res),
				user = new User(res,req),
				utokens = new UTokens(res),
				todos = new Todos(res),
				notes = new Notes(res),
				calendar = new Calendar(res),
				school = new School();

			if (p === '/') {
				if (user.verify(cookie['UTOKEN'])) {
					term(res,302,{'Location':'/usr'});
				}
				else {
					term(res,200,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/public/index/index.xhtml`,'r'));
				}
			}
			else if (p.split('/')[1] === 'sign-up') {
				term(res,200,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/public/index/sign-up/index.xhtml`,'r'));
			}
			else if (p.split('/')[1] === 'privacy') {
				term(res,200,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/public/index/privacy/index.xhtml`,'r'));
			}
			else if (p.split('/')[1] === 'cookies') {
				term(res,200,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/public/index/cookies/index.xhtml`,'r'));
			}
			else if (p.split('/').includes('usr')) {
				if (p.split('/')[1] === 'usr') {
					if (req.method === 'POST') {
						let body;

						req.on('data',chunk => body = chunk.toString());
						req.on('end',() => {
							let post = qs.parse(body);

							if (post.type === 'sign-in') {
								if (post.usr && post.psw) {
									if (
										user.verify({
											usr: post.usr.toString(),
											psw: post.psw.toString()
										})
									) {
										if (!cookie['UTOKEN']) {
											if (user.request('usr',post.usr).utoken) {
												term(
													res,302,
													{
														'Set-Cookie': `UTOKEN=${user.request('usr',post.usr).utoken}; Path=/;`,
														'Location': '/usr'
													}
												);
											}
											else {
												utokens.create(post.usr,true);
											}
										}
										else {
											term(res,302,{'Location':'/usr'});
										}
									}
								}
								term(res,302,{'Location':'/'});
							}
							else if (post.type === 'sign-up') {
								if (post.usr && post.psw && post.rpsw) {
									if (!user.check(post.usr) && post.psw === post.rpsw) user.create(post);
									else term(res,302,{'Location':'/sign-up'});
								}
							}
							else {
								user.login(false);
							}
						});
					}
					else if (req.method === 'GET') {
						if (p.split('/')[2] === 'sign-out') {
							term(res,302,{'Set-Cookie':'UTOKEN=; Path=/;','Location':'/'});
						}
						else if (cookie['UTOKEN']) {
							if (user.verify(cookie['UTOKEN'])) {
								user.login(true);
							}
							else {
								term(res,302,{'Location':'/'});
							}
						}
						else term(res,302,{'Location':'/'});
					}
					else term(res,405,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/server/client/err/405.xhtml`,'r'));
				}
			}
			else if (p.split('/').includes('root')) {
				res.end();
			}
			else {
				const h = path.resolve(dir,p.replace(/^\/*/,'public/'));
				if (ex(h)) {
					fs.readFile(h,(err,dat) => {
						if (err) term(res,500,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/server/client/err/500.xhtml`,'r'));
						res.statusCode = 200;
						return res.end(dat);
					});
				}
				else term(res,404,{'Content-Type':'application/xhtml+xml'},render(res,req,`${dir}/server/client/err/404.xhtml`,'r'));
			}
		}).listen(port);
	};

const // crypt variables
	algorithm = 'aes-256-ctr',
	iv = crypto.randomBytes(16);

const // crypt functions
	encrypt = (t,k) => {
		const
			cipher = crypto.createCipheriv(algorithm,k,iv),
			encrypted = Buffer.concat([cipher.update(t),cipher.final()]);
		return {
			iv: iv.toString('hex'),
			content: encrypted.toString('hex')
		};
	},
	decrypt = (h,k) => {
		const
			decipher = crypto.createDecipheriv(algorithm,k,Buffer.from(h.iv,'hex')),
			decrypted = Buffer.concat([decipher.update(Buffer.from(h.content,'hex')),decipher.final()]);
		return decrypted.toString();
	};

const // security
	getuuid = n => {
		let str = '';
		for (var i = 0; i < n; i++) str += uuidv4().toString().replace(/\-/g,'');
		return str;
	},
	token = getuuid(Math.ceil(config.server.security.token_length / 32)),
	recover = getuuid(Math.ceil(config.server.security.recover_token_length / 32));

const // auth
	keys = t => {
		let h = `${dirs.dat}/auth/root/keys.hash`;

		if (ex(h)) {
			try {
				return JSON.parse(rd(`${dirs.dat}/auth/root/keys.hash`))[t];
			}
			catch (e) {
				cli.err(`There was an error retrieving root account credentials:\n\n${e}`);
				process.exit(1);
			}
		}
	},
	guid = () => {
		let str = rd(`${dirs.dat}/auth/guid.asc`);

		if (str.split('').length !== 32) {
			let nguuid = getuuid(1);
			fs.writeFile(`${dirs.dat}/auth/guid.asc`,nguuid,err => {
				if (err) throw err;
			});
			return nguuid;
		}
		return str;
	};

if (rd(`${dirs.usr}/${config.app.alt_name}-data/server/init`)) {
	try {
		(async () => {
			await dbclient.connect();
		})();
		db = dbclient.db(config.app.alt_name);
		start(config.server.port);
		cli.cls();
		cli.log(`Server running at https://${ip.address()}:${config.server.port}\n\n`);
	}
	catch (e) {
		cli.err(`The server could not be initiated:\n\n${e}`);
		process.exit(1);
	}
}
else {
	cli.err(
		'You cannot initiate the server without executing configure first!',
		`Set your current working directory to the root directory of this project, and run the following:\n\n${'./configure'.bold}\n\nAfter that, you can initiate the server.`
	);
	process.exit(1);
}